microsoft/terminal
# Accessibility: Set-up UIA Tree

## Summary of the Pull Request
I touched A LOT of files. The main reason for this is that the `WindowUiaProvider` was moved to `Microsoft::Console::Types` so that it can be accessed by both projects. This meant we had to make a lot of propagating changes. But more info on that later.

**The Basics of Accessibility**
- [What is a User Interaction Automation (UIA) Tree?](https://docs.microsoft.com/en-us/dotnet/framework/ui-automation/ui-automation-tree-overview)
- Other projects (i.e.: Narrator) can take advantage of this UIA tree and are used to present information within it.
- Some things like XAML already have a UIA Tree. So some UIA tree navigation and features are already there. It's just a matter of getting them hooked up and looking right.

**Accessibility in our Project**
There's a few important classes...
regarding Accessibility...
- **WindowUiaProvider**: This sets up the UIA tree for a window. So this is the top-level for the UIA tree.
- **ScreenInfoUiaProvider**: This sets up the UIA tree for a terminal buffer.
- **UiaTextRange**: This is essential to interacting with the UIA tree for the terminal buffer. Actually gets portions of the buffer and presents them.

regarding the Windows Terminal window...
- **BaseWindow**: The foundation to a window. Deals with HWNDs and that kind of stuff.
- **IslandWindow**: This extends `BaseWindow` and is actually what holds our Windows Terminal
- **NonClientIslandWindow**: An extension of the `IslandWindow`

regarding ConHost...
- **IConsoleWindow**: This is an interface for the console window.
- **Window**: This is the actual window for ConHost. Extends `IConsoleWindow`



 
## References


## PR Checklist
* [X] Closes #634


## Detailed Description of the Pull Request / Additional comments
- `IConsoleWindow` changes:
  - move into `Microsoft::Console::Types` (a shared space)
  - Have `IslandWindow` extend it
- `WindowUiaProvider` changes:
  - move into `Microsoft::Console::Types` (a shared space)
- Hook up `WindowUiaProvider` to IslandWindow (yay! we now have a tree)


## Validation Steps Performed
I used "inspect.exe" to verify that the tree is hooked up. Verified this on ConHost and Windows Terminal.

## Comments after merging in PR #1915
Please refer to #1915 for a large chunk of the work done here. There are also a few known issues/TODOs that carry over:
- [ ] ConHost: the UIA Rects for the caption buttons are missing. Still investigating this issue.
- [ ] Issue #1914 - Tracing needs to be reattached
- [ ] Issue #1352 - Windows Terminal: the ScreenInfoUiaProvider needs to be attached to the TermControl.
- [ ] #1993 
- [ ] #1992
